torch/_overrides.py

```diff
@@ -0,0 +1,242 @@
+             len(implementation_spec.defaults) !=
+             len(dispatcher_spec.defaults))):
+        raise RuntimeError('implementation and dispatcher for %s have '
+                           'different function signatures' % implementation)
+
+    if implementation_spec.defaults is not None:
+        if dispatcher_spec.defaults != (None,) * len(dispatcher_spec.defaults):
+            raise RuntimeError('dispatcher functions can only use None for '
+                               'default argument values')
+
+
+_wrapped_func_source = textwrap.dedent("""
+    @functools.wraps(implementation)
+    def {name}(*args, **kwargs):
+        relevant_args = dispatcher(*args, **kwargs)
+        return implement_torch_function(
+            implementation, {name}, relevant_args, args, kwargs)
+    """)
+
+
+def torch_function_dispatch(dispatcher, module=None, verify=True,
+                            docs_from_dispatcher=False):
+    """Decorator for adding dispatch with the __torch_function__ protocol.
+
+    See for example ``torch/functional.py`` for usage examples.
+
+    Parameters
+    ----------
+    dispatcher : callable
+        Function that when called like ``dispatcher(*args, **kwargs)`` with
+        arguments from the NumPy function call returns an iterable of
+        array-like arguments to check for ``__torch_function__``.
+    module : str, optional
+        ``__module__`` attribute to set on new function, e.g.,
+        ``module='torch'``.  By default, module is copied from the decorated
+        function.
+    verify : bool, optional
+        If True, verify the that the signature of the dispatcher and decorated
+        function signatures match exactly: all required and optional arguments
+        should appear in order with the same names, but the default values for
+        all optional arguments should be ``None``. Only disable verification
+        if the dispatcher's signature needs to deviate for some particular
+        reason, e.g., because the function has a signature like
+        ``func(*args, **kwargs)``.
+    docs_from_dispatcher : bool, optional
+        If True, copy docs from the dispatcher function onto the dispatched
+        function, rather than from the implementation. This is useful for
+        functions defined in C, which otherwise don't have docstrings.
+
+    Returns
+    -------
+    dispatcher : callable
+        Function suitable for decorating the implementation of a NumPy
+        function.
+
+    Notes
+    -----
+    The dispatcher should normally return a tuple containing all input
+    arguments that may have a ``__torch_function__`` attribute.
+
+    In some cases where that's not easily possible, e.g. ``torch.cat``, it is
+    also valid (if a little slower) to make the dispatcher function a generator
+    (i.e. use ``yield`` to return arguments one by one).
+
+    """
+    def decorator(implementation):
+        if verify:
+            verify_matching_signatures(implementation, dispatcher)
```

Comment:
Nice, I would not have thought to implement this.

Reply:
Thanks. I can't take the credit for that one though, stolen from NumPy:)